sridha2412/assignment3/wordsearch.py

```diff
@@ -0,0 +1,126 @@
+        """ Constructor to initlalize a dictionary """
+        self.root = TrieNode()
+        for word in words:
+            self.add_word(word)
+
+    def add_word(self, word: str):
+        """ Method to add a word to dictionary if not already present """
+        current = self.root
+        for char in word:
+            if char not in current.children:
+                current.children[char] = TrieNode()
+            current = current.children[char]
+        current.is_word = True
+
+    def is_prefix(self, prefix: str) -> bool:
+        """ Method to return true if prefix found in dictionary """
+        current = self.root
+        for char in prefix:
+            if char not in current.children:
+                return False
+            current = current.children[char]
+        return True
+
+    def is_word(self, word: str) -> bool:
+        """ Method to return true if word found in dictionary """
+        current = self.root
+        for char in word:
+            if char not in current.children:
+                return False
+            current = current.children[char]
+        return current.is_word
+
+
+class Grid:
+    def __init__(self, grid: List):
+        """ Constructor to initlalize a grid """
+        self.grid = grid
+
+    def _get_neighbours(self, x: int, y: int) -> List:
+        """ Method to find all possible neighbouring cells given position (x, y) in a grid """
+        # 8 possible positions - vertical, horizontal and diagonal
+        positions = [(x - 1, y - 1), (x - 1, y), (x - 1, y + 1), (x, y + 1), (x + 1, y + 1), (x + 1, y), (x + 1, y - 1), (x, y - 1)]
+        return [pos for pos in positions if 0 <= pos[0] < len(self.grid) and 0 <= pos[1] < len(self.grid[0])]
+
+    def _get_word(self, positions: List) -> str:
+        """ Method to find corresponding word given a list of positions in grid """
+        return ''.join(self.grid[pos[0]][pos[1]] for pos in positions)
+
+    def _find_words(self, dictionary: Dictionary, position: List) -> List[str]:
+        """ Method to return all words from given position """
+        words = []
+
+        word = self._get_word(position)
+        if (not dictionary.is_prefix(word)):
+            return []
+        if (dictionary.is_word(word)):
+            words.append(word)
+
+        neighbours = self._get_neighbours(position[-1][0], position[-1][1])
+        for pos in neighbours:
+            if pos not in position:
+                position.append(pos)
+                words.extend(self._find_words(dictionary, position))
+                position.remove(pos)
+
+        return words
+
+    def find_all_words(self, dictionary: Dictionary) -> List[str]:
+        """ Method to return List of all words found in grid """
+        words = []
+        for i in range(len(self.grid)):
+            for j in range(len(self.grid[i])):
+                for word in self._find_words(dictionary, [(i, j)]):
+                    if word not in words:
```

Comment:
Be careful to use a set for this use case: the complexity of "in" on a list is problematic: for each word found, the whole list is checked.
Remember that sets are really cool, and although "in" is very readable, the complexity depends on the datastructure, and on the implementation of the __contains__ method.

Reply:
thanks JD, I have pushed the new changes